cli/exchange/agbot.go

```diff
@@ -41,50 +42,64 @@ func AgbotList(org string, userPw string, agbot string, namesOnly bool) {
 	}
 }
 
-func formPatternId(patternOrg, pattern string) string {
-	return patternOrg + "_" + pattern
+func formPatternId(patternOrg, pattern, nodeOrg string) string {
+	return patternOrg + "_" + pattern + "_" + nodeOrg
 }
 
 type ExchangeAgbotPatterns struct {
 	Patterns map[string]interface{} `json:"patterns"`
 }
 
-func AgbotListPatterns(org, userPw, agbot, patternOrg, pattern string) {
+func AgbotListPatterns(org, userPw, agbot, patternOrg, pattern, nodeOrg string) {
 	cliutils.SetWhetherUsingApiKey(userPw)
 	org, agbot = cliutils.TrimOrg(org, agbot)
 	var patternId string
 	if patternOrg != "" || pattern != "" {
 		if patternOrg == "" || pattern == "" {
 			cliutils.Fatal(cliutils.CLI_INPUT_ERROR, "both patternorg and pattern must be specified (or neither)")
 		}
-		patternId = formPatternId(patternOrg, pattern)
+		if nodeOrg == "" {
+			nodeOrg = patternOrg
+		}
+		patternId = formPatternId(patternOrg, pattern, nodeOrg)
 	}
 	// Display the full resources
 	var patterns ExchangeAgbotPatterns
 	httpCode := cliutils.ExchangeGet(cliutils.GetExchangeUrl(), "orgs/"+org+"/agbots/"+agbot+"/patterns"+cliutils.AddSlash(patternId), cliutils.OrgAndCreds(org, userPw), []int{200, 404}, &patterns)
 	if httpCode == 404 && patternOrg != "" && pattern != "" {
-		cliutils.Fatal(cliutils.NOT_FOUND, "pattern '%s' with org '%s' not found in agbot '%s'", pattern, patternOrg, agbot)
+		cliutils.Fatal(cliutils.NOT_FOUND, "pattern '%s' with org '%s' and node org '%s' not found in agbot '%s'", pattern, patternOrg, nodeOrg, agbot)
 	}
-	output := cliutils.MarshalIndent(patterns.Patterns, "exchange agbot patterns list")
+	output := cliutils.MarshalIndent(patterns.Patterns, "exchange agbot listpattern")
 	fmt.Println(output)
 }
 
 type ServedPattern struct {
-	Org     string `json:"patternOrgid"`
+	PatternOrg     string `json:"patternOrgid"`
 	Pattern string `json:"pattern"`
+	NodeOrg     string `json:"nodeOrgid"`
 }
 
-func AgbotAddPattern(org, userPw, agbot, patternOrg, pattern string) {
+func AgbotAddPattern(org, userPw, agbot, patternOrg, pattern, nodeOrg string) {
 	cliutils.SetWhetherUsingApiKey(userPw)
 	org, agbot = cliutils.TrimOrg(org, agbot)
-	patternId := formPatternId(patternOrg, pattern)
-	input := ServedPattern{Org: patternOrg, Pattern: pattern}
-	cliutils.ExchangePutPost(http.MethodPut, cliutils.GetExchangeUrl(), "orgs/"+org+"/agbots/"+agbot+"/patterns/"+patternId, cliutils.OrgAndCreds(org, userPw), []int{201}, input)
+	if nodeOrg == "" {
+		nodeOrg = patternOrg
+	}
+	//patternId := formPatternId(patternOrg, pattern, nodeOrg)
+	input := ServedPattern{PatternOrg: patternOrg, Pattern: pattern, NodeOrg: nodeOrg}
```

Comment:
With this change, do we need to increase the `MINIMUM_EXCHANGE_VERSION` and `PREFERRED_EXCHANGE_VERSION` in version/version.go?

Reply:
Yes, i think so. Can you do that in your PR when you exploit the new exchange api's?

Reply:
yes